File: sdks/java/src/main/java/com/legacybaas/sdk/LegacyBaaSClient.java
/**
 * LegacyBAAS Java Client
 * Main client for all SDK operations
 */

package com.legacybaas.sdk;

import com.legacybaas.sdk.errors.*;
import com.legacybaas.sdk.services.*;
import com.legacybaas.sdk.utils.Logger;
import com.legacybaas.sdk.utils.RetryHandler;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.JsonNode;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.HashMap;
import java.util.concurrent.CompletableFuture;

/**
 * Main LegacyBAAS client for Java SDK
 */
public class LegacyBaaSClient {
    private final String clientId;
    private final String clientSecret;
    private final String baseUrl;
    private final String environment;
    private final Duration timeout;
    private final int maxRetries;
    private final Duration retryDelay;
    private final Logger logger;
    
    private String accessToken;
    private Instant tokenExpiry;
    
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final RetryHandler retryHandler;
    
    // Services
    private final SwiftProcessor swiftProcessor;
    private final BlockchainRouter blockchainRouter;
    private final BankingService bankingService;
    private final AnalyticsService analyticsService;
    private final WebhookHandler webhookHandler;
    private final ComplianceService complianceService;
    
    /**
     * Constructor
     */
    public LegacyBaaSClient(Builder builder) {
        this.clientId = builder.clientId;
        this.clientSecret = builder.clientSecret;
        this.baseUrl = builder.baseUrl != null ? builder.baseUrl : getBaseUrl(builder.environment);
        this.environment = builder.environment;
        this.timeout = builder.timeout;
        this.maxRetries = builder.maxRetries;
        this.retryDelay = builder.retryDelay;
        this.logger = builder.enableLogging ? new Logger(builder.logLevel) : null;
        
        // Initialize HTTP client
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(timeout)
            .build();
        
        this.objectMapper = new ObjectMapper();
        this.retryHandler = new RetryHandler(maxRetries, retryDelay);
        
        // Initialize services
        this.swiftProcessor = new SwiftProcessor(this);
        this.blockchainRouter = new BlockchainRouter(this);
        this.bankingService = new BankingService(this);
        this.analyticsService = new AnalyticsService(this);
        this.webhookHandler = new WebhookHandler(this);
        this.complianceService = new ComplianceService(this);
        
        if (logger != null) {
            logger.info("LegacyBAAS Client initialized - Environment: " + environment + ", Base URL: " + baseUrl);
        }
    }
    
    /**
     * Builder class for LegacyBaaSClient
     */
    public static class Builder {
        private String clientId;
        private String clientSecret;
        private String baseUrl;
        private String environment = "production";
        private Duration timeout = Duration.ofSeconds(30);
        private int maxRetries = 3;
        private Duration retryDelay = Duration.ofSeconds(1);
        private boolean enableLogging = true;
        private String logLevel = "INFO";
        
        public Builder clientId(String clientId) {
            this.clientId = clientId;
            return this;
        }
        
        public Builder clientSecret(String clientSecret) {
            this.clientSecret = clientSecret;
            return this;
        }
        
        public Builder baseUrl(String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }
        
        public Builder environment(String environment) {
            this.environment = environment;
            return this;
        }
        
        public Builder timeout(Duration timeout) {
            this.timeout = timeout;
            return this;
        }
        
        public Builder maxRetries(int maxRetries) {
            this.maxRetries = maxRetries;
            return this;
        }
        
        public Builder retryDelay(Duration retryDelay) {
            this.retryDelay = retryDelay;
            return this;
        }
        
        public Builder enableLogging(boolean enableLogging) {
            this.enableLogging = enableLogging;
            return this;
        }
        
        public Builder logLevel(String logLevel) {
            this.logLevel = logLevel;
            return this;
        }
        
        public LegacyBaaSClient build() {
            if (clientId == null || clientSecret == null) {
                throw new IllegalArgumentException("Client ID and Client Secret are required");
            }
            return new LegacyBaaSClient(this);
        }
    }
    
    /**
     * Get base URL for environment
     */
    private String getBaseUrl(String environment) {
        switch (environment.toLowerCase()) {
            case "production":
                return "https://api.legacybaas.com/v1";
            case "sandbox":
                return "https://sandbox.legacybaas.com/v1";
            case "development":
                return "http://localhost:3000/v1";
            default:
                return "https://api.legacybaas.com/v1";
        }
    }
    
    /**
     * Authenticate with OAuth2 client credentials flow
     */
    public CompletableFuture<Map<String, Object>> authenticate() {
        if (clientId == null || clientSecret == null) {
            return CompletableFuture.failedFuture(
                new AuthenticationError("Client ID and Client Secret are required for authentication")
            );
        }
        
        Map<String, Object> data = new HashMap<>();
        data.put("grant_type", "client_credentials");
        data.put("client_id", clientId);
        data.put("client_secret", clientSecret);
        data.put("scope", "banking:read banking:write blockchain:execute compliance:screen");
        
        return makeRequest("POST", "/auth/token", data, true)
            .thenApply(response -> {
                Map<String, Object> tokenData = (Map<String, Object>) response;
                this.accessToken = (String) tokenData.get("access_token");
                Integer expiresIn = (Integer) tokenData.get("expires_in");
                this.tokenExpiry = Instant.now().plusSeconds(expiresIn);
                
                if (logger != null) {
                    logger.info("Authentication successful");
                }
                
                return tokenData;
            })
            .exceptionally(throwable -> {
                if (logger != null) {
                    logger.error("Authentication failed: " + throwable.getMessage());
                }
                throw new AuthenticationError("Failed to authenticate with LegacyBAAS API");
            });
    }
    
    /**
     * Check if client has valid authentication token
     */
    public boolean isAuthenticated() {
        return accessToken != null && tokenExpiry != null && Instant.now().isBefore(tokenExpiry);
    }
    
    /**
     * Ensure client is authenticated
     */
    public CompletableFuture<Void> ensureAuthenticated() {
        if (!isAuthenticated()) {
            return authenticate().thenAccept(response -> {});
        }
        return CompletableFuture.completedFuture(null);
    }
    
    /**
     * Make HTTP request to API
     */
    public CompletableFuture<Object> makeRequest(String method, String endpoint, Object data, boolean skipAuth) {
        CompletableFuture<Void> authFuture = skipAuth ? 
            CompletableFuture.completedFuture(null) : 
            ensureAuthenticated();
            
        return authFuture.thenCompose(v -> {
            try {
                String url = baseUrl + endpoint;
                
                HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .timeout(timeout)
                    .header("Content-Type", "application/json")
                    .header("User-Agent", "LegacyBAAS-Java-SDK/2.0.0")
                    .header("Accept", "application/json");
                
                if (accessToken != null && !skipAuth) {
                    requestBuilder.header("Authorization", "Bearer " + accessToken);
                }
                
                HttpRequest request;
                switch (method.toUpperCase()) {
                    case "GET":
                        request = requestBuilder.GET().build();
                        break;
                    case "POST":
                        String jsonData = data != null ? objectMapper.writeValueAsString(data) : "";
                        request = requestBuilder.POST(HttpRequest.BodyPublishers.ofString(jsonData)).build();
                        break;
                    case "PUT":
                        String putData = data != null ? objectMapper.writeValueAsString(data) : "";
                        request = requestBuilder.PUT(HttpRequest.BodyPublishers.ofString(putData)).build();
                        break;
                    case "DELETE":
                        request = requestBuilder.DELETE().build();
                        break;
                    default:
                        throw new IllegalArgumentException("Unsupported HTTP method: " + method);
                }
                
                if (logger != null) {
                    logger.debug("Making " + method + " request to " + endpoint);
                }
                
                return httpClient.sendAsync(request, HttpResponse.BodyHandlers.ofString())
                    .thenApply(this::handleResponse);
                    
            } catch (Exception e) {
                return CompletableFuture.failedFuture(new NetworkError("Request failed: " + e.getMessage()));
            }
        });
    }
    
    /**
     * Handle HTTP response
     */
    private Object handleResponse(HttpResponse<String> response) {
        try {
            if (response.statusCode() >= 400) {
                JsonNode errorData;
                try {
                    errorData = objectMapper.readTree(response.body());
                } catch (Exception e) {
                    errorData = objectMapper.createObjectNode().put("message", response.body());
                }
                
                String message = errorData.has("message") ? errorData.get("message").asText() : "HTTP " + response.statusCode();
                
                switch (response.statusCode()) {
                    case 401:
                        throw new AuthenticationError(message);
                    case 400:
                        throw new ValidationError(message);
                    case 429:
                        throw new RateLimitError(message);
                    case 500:
                    case 502:
                    case 503:
                    case 504:
                        throw new NetworkError(message);
                    default:
                        throw new LegacyBaaSError(message, response.statusCode());
                }
            }
            
            if (logger != null) {
                logger.debug("Request completed: " + response.request().method() + " " + response.request().uri());
            }
            
            // Parse response
            JsonNode responseData = objectMapper.readTree(response.body());
            if (responseData.has("data")) {
                return objectMapper.convertValue(responseData.get("data"), Object.class);
            }
            return objectMapper.convertValue(responseData, Object.class);
            
        } catch (IOException e) {
            throw new NetworkError("Failed to parse response: " + e.getMessage());
        }
    }
    
    // Service getters
    public SwiftProcessor getSwiftProcessor() { return swiftProcessor; }
    public BlockchainRouter getBlockchainRouter() { return blockchainRouter; }
    public BankingService getBankingService() { return bankingService; }
    public AnalyticsService getAnalyticsService() { return analyticsService; }
    public WebhookHandler getWebhookHandler() { return webhookHandler; }
    public ComplianceService getComplianceService() { return complianceService; }
    
    /**
     * Get API health status
     */
    public CompletableFuture<Object> getHealth() {
        return makeRequest("GET", "/health", null, false);
    }
    
    /**
     * Get platform metrics
     */
    public CompletableFuture<Object> getMetrics() {
        return makeRequest("GET", "/admin/metrics", null, false);
    }
    
    /**
     * Close client and cleanup resources
     */
    public void close() {
        this.accessToken = null;
        this.tokenExpiry = null;
        
        if (logger != null) {
            logger.info("LegacyBAAS Client closed");
        }
    }
}

File: sdks/typescript/src/client/LegacyBaaSClient.ts
/**
 * LegacyBAAS Main Client
 * Entry point for all SDK operations
 */

import { EventEmitter } from 'events';
import { SwiftProcessor } from '../services/SwiftProcessor';
import { BlockchainRouter } from '../services/BlockchainRouter';
import { BankingService } from '../services/BankingService';
import { AnalyticsService } from '../services/AnalyticsService';
import { WebhookHandler } from '../services/WebhookHandler';
import { ComplianceService } from '../services/ComplianceService';
import { ClientConfig, ApiResponse } from '../types';
import { AuthenticationError, LegacyBaaSError } from '../errors';
import { Logger } from '../utils/Logger';
import { RetryHandler } from '../utils/RetryHandler';

export interface LegacyBaaSClientConfig {
  apiKey?: string;
  clientId?: string;
  clientSecret?: string;
  baseUrl?: string;
  environment?: 'production' | 'sandbox' | 'development';
  timeout?: number;
  maxRetries?: number;
  retryDelay?: number;
  enableRealTime?: boolean;
  enableLogging?: boolean;
  logLevel?: 'DEBUG' | 'INFO' | 'WARN' | 'ERROR';
}

export class LegacyBaaSClient extends EventEmitter {
  private config: ClientConfig;
  private accessToken?: string;
  private tokenExpiry?: number;
  private logger: Logger;
  private retryHandler: RetryHandler;
  
  // Service instances
  public readonly swift: SwiftProcessor;
  public readonly blockchain: BlockchainRouter;
  public readonly banking: BankingService;
  public readonly analytics: AnalyticsService;
  public readonly webhooks: WebhookHandler;
  public readonly compliance: ComplianceService;

  constructor(config: LegacyBaaSClientConfig) {
    super();
    
    // Set default configuration
    this.config = {
      baseUrl: this.getBaseUrl(config.environment || 'production'),
      version: 'v1',
      timeout: config.timeout || 30000,
      maxRetries: config.maxRetries || 3,
      retryDelay: config.retryDelay || 1000,
      enableRealTime: config.enableRealTime !== false,
      enableLogging: config.enableLogging !== false,
      logLevel: config.logLevel || 'INFO',
      ...config
    };

    // Initialize utilities
    this.logger = new Logger(this.config.logLevel);
    this.retryHandler = new RetryHandler(this.config);

    // Initialize services
    this.swift = new SwiftProcessor(this);
    this.blockchain = new BlockchainRouter(this);
    this.banking = new BankingService(this);
    this.analytics = new AnalyticsService(this);
    this.webhooks = new WebhookHandler(this);
    this.compliance = new ComplianceService(this);

    this.logger.info('LegacyBAAS Client initialized', {
      environment: config.environment,
      baseUrl: this.config.baseUrl,
      enableRealTime: this.config.enableRealTime
    });
  }

  private getBaseUrl(environment: string): string {
    switch (environment) {
      case 'production':
        return 'https://api.legacybaas.com/v1';
      case 'sandbox':
        return 'https://sandbox.legacybaas.com/v1';
      case 'development':
        return 'http://localhost:3000/v1';
      default:
        return 'https://api.legacybaas.com/v1';
    }
  }

  /**
   * Authenticate with the LegacyBAAS API
   */
  public async authenticate(): Promise<void> {
    if (!this.config.clientId || !this.config.clientSecret) {
      throw new AuthenticationError('Client ID and Client Secret are required for authentication');
    }

    try {
      const response = await this.makeRequest<{
        access_token: string;
        token_type: string;
        expires_in: number;
      }>('POST', '/auth/token', {
        grant_type: 'client_credentials',
        client_id: this.config.clientId,
        client_secret: this.config.clientSecret,
        scope: 'banking:read banking:write blockchain:execute compliance:screen'
      }, { skipAuth: true });

      this.accessToken = response.data.access_token;
      this.tokenExpiry = Date.now() + (response.data.expires_in * 1000);
      
      this.logger.info('Authentication successful');
      this.emit('authenticated', { expiresIn: response.data.expires_in });
      
    } catch (error) {
      this.logger.error('Authentication failed', error);
      throw new AuthenticationError('Failed to authenticate with LegacyBAAS API');
    }
  }

  /**
   * Check if the current token is valid
   */
  public isAuthenticated(): boolean {
    return !!(this.accessToken && this.tokenExpiry && Date.now() < this.tokenExpiry);
  }

  /**
   * Ensure we have a valid token, refresh if necessary
   */
  public async ensureAuthenticated(): Promise<void> {
    if (!this.isAuthenticated()) {
      await this.authenticate();
    }
  }

  /**
   * Make HTTP request to the API
   */
  public async makeRequest<T>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    endpoint: string,
    data?: any,
    options: { skipAuth?: boolean; headers?: Record<string, string> } = {}
  ): Promise<ApiResponse<T>> {
    
    if (!options.skipAuth) {
      await this.ensureAuthenticated();
    }

    const url = `${this.config.baseUrl}${endpoint}`;
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'User-Agent': 'LegacyBAAS-SDK-TS/2.0.0',
      ...options.headers
    };

    if (this.accessToken && !options.skipAuth) {
      headers.Authorization = `Bearer ${this.accessToken}`;
    }

    const requestOptions: RequestInit = {
      method,
      headers,
      signal: AbortSignal.timeout(this.config.timeout)
    };

    if (data && (method === 'POST' || method === 'PUT')) {
      requestOptions.body = JSON.stringify(data);
    }

    try {
      const response = await this.retryHandler.retry(async () => {
        this.logger.debug(`Making ${method} request to ${endpoint}`, { data });
        
        const fetchResponse = await fetch(url, requestOptions);
        
        if (!fetchResponse.ok) {
          const errorData = await fetchResponse.json().catch(() => ({}));
          throw new LegacyBaaSError(
            errorData.message || `HTTP ${fetchResponse.status}`,
            fetchResponse.status,
            errorData.code
          );
        }

        return fetchResponse.json();
      });

      this.logger.debug(`Request completed: ${method} ${endpoint}`, { 
        status: response.success ? 'success' : 'error' 
      });

      return response;
      
    } catch (error) {
      this.logger.error(`Request failed: ${method} ${endpoint}`, error);
      throw error;
    }
  }

  /**
   * Get client configuration
   */
  public getConfig(): ClientConfig {
    return { ...this.config };
  }

  /**
   * Update client configuration
   */
  public updateConfig(updates: Partial<LegacyBaaSClientConfig>): void {
    this.config = { ...this.config, ...updates };
    this.logger.info('Client configuration updated', updates);
  }

  /**
   * Get API health status
   */
  public async getHealth(): Promise<ApiResponse<{
    status: string;
    version: string;
    timestamp: string;
    services: Record<string, string>;
  }>> {
    return this.makeRequest('GET', '/health');
  }

  /**
   * Get platform metrics
   */
  public async getMetrics(): Promise<ApiResponse<{
    system: Record<string, number>;
    api: Record<string, number>;
    banking: Record<string, number>;
    blockchain: Record<string, number>;
  }>> {
    return this.makeRequest('GET', '/admin/metrics');
  }

  /**
   * Close client and cleanup resources
   */
  public async close(): Promise<void> {
    this.removeAllListeners();
    this.accessToken = undefined;
    this.tokenExpiry = undefined;
    this.logger.info('LegacyBAAS Client closed');
  }
}